conn_http_batch.go

```diff
@@ -0,0 +1,201 @@
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. ClickHouse, Inc. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package clickhouse
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/ClickHouse/clickhouse-go/v2/lib/binary"
+	"github.com/ClickHouse/clickhouse-go/v2/lib/column"
+	"github.com/ClickHouse/clickhouse-go/v2/lib/driver"
+	"github.com/ClickHouse/clickhouse-go/v2/lib/proto"
+	"io"
+	"io/ioutil"
+	"os"
+	"regexp"
+	"strings"
+)
+
+var splitHttpInsertRe = regexp.MustCompile(`(?i)INSERT INTO\s([\w.]+)(\s\((.*?)\))?`)
+
+// release is ignored, because http used by std with empty release function
+func (h *httpConnect) prepareBatch(ctx context.Context, query string, release func(*connect, error)) (driver.Batch, error) {
+	index := splitHttpInsertRe.FindStringSubmatchIndex(strings.ToUpper(query))
+
+	if len(index) < 3 {
+		return nil, errors.New("cannot get table name from query")
+	}
+
+	tableName := query[index[2]:index[3]]
+	query = "INSERT INTO " + tableName + " FORMAT Native"
+	queryTableSchema := "DESCRIBE TABLE " + tableName
+	r, err := h.query(ctx, release, queryTableSchema)
+	if err != nil {
+		return nil, err
+	}
+
+	block := &proto.Block{}
+
+	// get Table columns and types
+	for r.Next() {
+		var (
+			colName string
+			colType string
+			ignore  string
+		)
+
+		err = r.Scan(&colName, &colType, &ignore, &ignore, &ignore, &ignore, &ignore)
+		if err != nil {
+			return nil, err
+		}
+
+		err = block.AddColumn(colName, column.Type(colType))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &httpBatch{
+		ctx:       ctx,
+		conn:      h,
+		structMap: &structMap{},
+		block:     block,
+		query:     query,
+	}, nil
+}
+
+type httpBatch struct {
+	query     string
+	err       error
+	ctx       context.Context
+	conn      *httpConnect
+	structMap *structMap
+	sent      bool
+	block     *proto.Block
+}
+
+func (b *httpBatch) Abort() error {
+	defer func() {
+		b.sent = true
+	}()
+	if b.sent {
+		return ErrBatchAlreadySent
+	}
+	return nil
+}
+
+func (b *httpBatch) Append(v ...interface{}) error {
+	if b.sent {
+		return ErrBatchAlreadySent
+	}
+	if err := b.block.Append(v...); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (b *httpBatch) AppendStruct(v interface{}) error {
+	values, err := b.structMap.Map("AppendStruct", b.block.ColumnsNames(), v, false)
```

Comment:
this is for later but use of DESCRIBE means we can extract default values and address https://github.com/ClickHouse/clickhouse-go/issues/587